### PR TITLE
attempt for #451

### LINKED ISSUE #451: Add `reward` and `observation` arguments to env.reset()
## 🚀 Feature

The [CompilerEnv constructor](https://facebookresearch.github.io/CompilerGym/compiler_gym/envs.html#compiler_gym.envs.CompilerEnv) accepts a pair of arguments `reward_space` and `observation_space`. We should add those to `env.reset()`, same as specifying the benchmark.

## Motivation

Because this feels like a clumsy API:

```py
env.reward_space = "Foobar"
env.reset(benchmark="benchmark://foo-v0/abc")
```

## Pitch

Allow:

```py
env.reset(benchmark="benchmark://foo-v0/abc", reward_space="Foobar")
```

### PR DESCRIPTION
**Pull request by SoumyajitKarmakar**
Modified the env.reset() function to add the reward_space and observation_space parameters. Modified the reset() func of the subclasses. Added a OptionalArgumentValue class in gym_type_hints.py for the default value.

Fixes #451

### CODE REVIEW

**Comment by ChrisCummins** [comment]
Hi @SoumyajitKarmakar, sounds great! Would you like me to review your changes or are you still working on this?

Cheers,
Chris

**Comment by SoumyajitKarmakar** [comment]
Hi @ChrisCummins,
Yes, please review my changes and let me know if it requires any additional modifications.

File: tests/compiler_env_test.py [hunk 1]
```diff
@@ -147,9 +147,9 @@ def test_switch_default_reward_space_in_episode(env: LlvmEnv):
     """Test that switching reward space during an episode resets the cumulative
     episode reward.
     """
-    env.reward_space = None
+    # env.reward_space = None
```
Comment (ChrisCummins) [inline tests/compiler_env_test.py]: Nitpick: Don't commit uncommented code, just remove the line

File: tests/compiler_env_test.py [hunk 2]
```diff
@@ -167,9 +167,9 @@ def test_set_same_default_reward_space_in_episode(env: LlvmEnv):
     """Test that setting the reward space during an episode does not reset the
     cumulative episode reward if the reward space is unchanged.
     """
-    env.reward_space = "IrInstructionCount"
+    # env.reward_space = "IrInstructionCount"
```
Comment (ChrisCummins) [inline tests/compiler_env_test.py]: Same again here

**Review by ChrisCummins — COMMENTED** [review]
Hi @SoumyajitKarmakar, thanks for your patch, your changes look good!

Since you are adding new features, could you please add unit tests to cover these new features? I think `tests/compiler_env_test.py` might be a good place to add them, and you can take a look at the existing tests for inspiration.

A couple of other minor nitpicks left inline. Once you push these changes, please ping me for another review.

Cheers,
Chris

**Comment by SoumyajitKarmakar** [comment]
Hi @ChrisCummins,
Please review the latest changes. I have removed the unnecessary commented lines and added some tests.
It is my first time writing tests for a software. I have added 4 simple tests which should hopefully cover the new features. Please let me know if any other kinds of tests are required.

File: tests/compiler_env_test.py [hunk 3]
```diff
@@ -34,6 +34,30 @@ def test_benchmark_set_in_reset(env: LlvmEnv):
     assert env.benchmark == "benchmark://cbench-v1/dijkstra"
 
 
+def test_reward_space_setter(env: LlvmEnv):
+    env.reward_space = "IrInstructionCount"
+    assert env.reward_space != "IrInstructionCount"
+    env.reset()
+    assert env.reward_space == "IrInstructionCount"
```
Comment (ChrisCummins) [inline tests/compiler_env_test.py]: For this and the other 4 assertions you added, you must convert `env.reward_space` and `env.observation_space` to strings to check their value:

```suggestion
    assert str(env.reward_space) == "IrInstructionCount"
```

**Review by ChrisCummins — COMMENTED** [review]
Thanks for adding the tests! The test logic looks good, you just need to make sure to convert the `env.reward_space` and `env.observation_space` attributes to strings for comparison.

Once you fix that, this looks ready to merge.

Cheers,
Chris

File: tests/compiler_env_test.py (hunk 3 above)
Comment (SoumyajitKarmakar) [reply tests/compiler_env_test.py]: Oh my bad, fixed them.

**Comment by ChrisCummins** [comment]
Thanks @SoumyajitKarmakar for fixing this :) This LGTM now, but holding off on merging until I fix #657.

Cheers,
Chris

**Comment by ChrisCummins** [comment]
Ah, looks like I gave you bad advice on how to fix the failing tests. Not your fault, I'll patch this up myself and merge via #659.

Thanks for contributing to CompilerGym @SoumyajitKarmakar! 🎉 

Cheers,
Chris

**Comment by SoumyajitKarmakar** [comment]
It was a pleasure working with you.

**Pull request closed without merging**